hazelcast/src/main/java/com/hazelcast/spi/discovery/impl/DefaultDiscoveryService.java

```diff
@@ -155,11 +155,17 @@ private boolean validateCandidate(DiscoveryNode candidate) {
 
             List<DiscoveryStrategy> discoveryStrategies = new ArrayList<DiscoveryStrategy>();
             for (DiscoveryStrategyFactory factory : factories) {
+                //as a side-effect the buildDiscoveryStrategy() removes a config of the created strategy
                 DiscoveryStrategy discoveryStrategy = buildDiscoveryStrategy(factory, discoveryStrategyConfigs);
                 if (discoveryStrategy != null) {
                     discoveryStrategies.add(discoveryStrategy);
                 }
             }
+            if (!discoveryStrategyConfigs.isEmpty()) {
```

Comment:
I think it should be quite easy to accomplish. Currently we iterate in this order:
```java
for (DiscoveryStrategyFactory factory : factories) {
   for (DiscoveryStrategyConfig config : discoveryStrategyConfigs) {
   }
}
```
If we switch those loops, we should be able to directly throw the exception, when no factory was found for a configuration (right now that `return null` in `buildDiscoveryStrategy()` is not an immediate error, since the next factory could be the matching one).